rbd/pool_nautilus.go

```diff
@@ -84,3 +84,151 @@ func RemovePoolMetadata(ioctx *rados.IOContext, key string) error {
+	// RBD_POOL_STAT_OPTION_IMAGES from librbd.
+	PoolStatOptionImages = PoolStatOption(C.RBD_POOL_STAT_OPTION_IMAGES)
+	// PoolStatOptionImageProvisionedBytes is the representation of
+	// RBD_POOL_STAT_OPTION_IMAGE_PROVISIONED_BYTES from librbd.
+	PoolStatOptionImageProvisionedBytes = PoolStatOption(C.RBD_POOL_STAT_OPTION_IMAGE_PROVISIONED_BYTES)
+	// PoolStatOptionImageMaxProvisionedBytes is the representation of
+	// RBD_POOL_STAT_OPTION_IMAGE_MAX_PROVISIONED_BYTES from librbd.
+	PoolStatOptionImageMaxProvisionedBytes = PoolStatOption(C.RBD_POOL_STAT_OPTION_IMAGE_MAX_PROVISIONED_BYTES)
+	// PoolStatOptionImageSnapshots is the representation of
+	// RBD_POOL_STAT_OPTION_IMAGE_SNAPSHOTS from librbd.
+	PoolStatOptionImageSnapshots = PoolStatOption(C.RBD_POOL_STAT_OPTION_IMAGE_SNAPSHOTS)
+	// PoolStatOptionTrashImages is the representation of
+	// RBD_POOL_STAT_OPTION_TRASH_IMAGES from librbd.
+	PoolStatOptionTrashImages = PoolStatOption(C.RBD_POOL_STAT_OPTION_TRASH_IMAGES)
+	// PoolStatOptionTrashProvisionedBytes is the representation of
+	// RBD_POOL_STAT_OPTION_TRASH_PROVISIONED_BYTES from librbd.
+	PoolStatOptionTrashProvisionedBytes = PoolStatOption(C.RBD_POOL_STAT_OPTION_TRASH_PROVISIONED_BYTES)
+	// PoolStatOptionTrashMaxProvisionedBytes is the representation of
+	// RBD_POOL_STAT_OPTION_TRASH_MAX_PROVISIONED_BYTES from librbd.
+	PoolStatOptionTrashMaxProvisionedBytes = PoolStatOption(C.RBD_POOL_STAT_OPTION_TRASH_MAX_PROVISIONED_BYTES)
+	// PoolStatOptionTrashSnapshots is the representation of
+	// RBD_POOL_STAT_OPTION_TRASH_SNAPSHOTS from librbd.
+	PoolStatOptionTrashSnapshots = PoolStatOption(C.RBD_POOL_STAT_OPTION_TRASH_SNAPSHOTS)
+)
+
+// addPoolStatOption adds the given PoolStatOption to PoolStats.
+//
+// Implements:
+//  int rbd_pool_stats_option_add_uint64(rbd_pool_stats_t stats, int stat_option, uint64_t* stat_val)
+func (poolStats *PoolStats) addPoolStatOption(option PoolStatOption, val *uint64) error {
+	ret := C.rbd_pool_stats_option_add_uint64(
+		poolStats.stats,
+		C.int(option),
+		(*C.uint64_t)(val))
+	return getError(ret)
+}
+
+// GetPoolStats returns the value for the PoolStatOption.
+func (poolStats *PoolStats) GetPoolStats(ioctx *rados.IOContext, option PoolStatOption) (uint64, error) {
+	var val uint64
+	if ioctx == nil {
+		return val, ErrNoIOContext
+	}
+
+	err := poolStats.addPoolStatOption(option, &val)
+	if err != nil {
+		return val, err
+	}
+
+	ret := C.rbd_pool_stats_get(cephIoctx(ioctx), poolStats.stats)
+	return val, getError(ret)
+}
+
+// GetAllPoolStats returns a map of all PoolStatOption(s) to their respective values.
+//
+// Implements:
+//  int rbd_pool_stats_get(cephIoctx(ioctx), poolStats.stats)
+func GetAllPoolStats(ioctx *rados.IOContext) (map[PoolStatOption]uint64, error) {
+	var omap = make(map[PoolStatOption]uint64)
+	if ioctx == nil {
+		return omap, ErrNoIOContext
+	}
+
+	poolStats := PoolStatsCreate()
+	defer func() {
+		poolStats.Destroy()
+	}()
+
+	var keys [8]PoolStatOption
```

Comment:
I don't understand the approach used here. Why not set up the map in the first place like:
```
m := map[PoolStatOption]uint64{
    PoolStatOptionImages: 0,
    PoolStatOPtionImageProvsionedBytes: 0,
    // ...and so on...
}
```

Reply:
I need to pass the address in addOption() and I can't pass the address of map value, so either I pass the address through different variables same like python code or I keep a slice for values and pass the address of the respective values so that value slice is named ovalArray.

Now, why I used a slice named keys is because I need to iterate in an order but during map iteration order of keys is not guranteed. I coded the suggested approach first but changed it because of the order issue.

Reply:
OK, that's a bit annoying but I understand now.